### PR TITLE
[RFC] Recursive mutexes

### PR DESCRIPTION
**Pull request by jforissier**
A couple of commits to add support for recursive mutexes, and use one in the mempool functions `get_pool()` and `put_pool()`.

I think it makes the code a bit easier to read, and it also saves one `mutex_unlock()` and one `mutex_lock()` call in a sequence: `get_pool(); get_pool(); ... put_pool(); put_pool();`. The reason is because the mutex remains locked during the whole sequence instead of being locked + unlocked in the first `get_pool()` and again in the last `put_pool()`. I don't expect it to make any practical difference though.

A small drawback is the size of the `mutex` structure is increased for all users. On QEMU that's an additional 104 bytes of `.bss` and another 112 bytes of `.text`, not much at all.

So I am posting this as a RFC. We could also defer until we have at least two users and not just mempool. I did not find another candidate at the moment.

### CODE REVIEW

**Comment by jforissier** [comment]
IBART is broken by https://github.com/OP-TEE/manifest/pull/165 :cry: 
@jbech-linaro  I suppose the "failed" messages stop the script with an error?
It seems I will have to look into that. I thought I could just ignore them... https://github.com/96boards-hikey/burn-boot/commit/a55c068f9a0aaf9da440aacdfaf85182e168e36c

**Comment by jforissier** [comment]
@jbech-linaro is there a way you could test this PR: https://github.com/96boards-hikey/burn-boot/pull/4 before it is merged and taken in our manifest? Thanks.

**Comment by jbech-linaro** [comment]
> @jbech-linaro is there a way you could test this PR: [96boards-hikey/burn-boot#4](https://github.com/96boards-hikey/burn-boot/pull/4) before it is merged and taken in our manifest? Thanks.

I think the easiest is to just try it out locally on a machine first. I had my HiKey up and running last week, so it should be quick and easy for me to give your patches a try. I'll try to do it tomorrow.

**Comment by jforissier** [comment]
I have done a few local tests myself, but more is better ;-) one never knows what can happen, different board variants etc. Thanks!

File: core/arch/arm/kernel/mutex.c [hunk 1]
```diff
@@ -45,6 +57,10 @@ static void __mutex_lock(struct mutex *m, const char *fname, int lineno)
 			wq_wait_init(&m->wq, &wqe, false /* wait_read */);
 		} else {
 			m->state = -1; /* write locked */
+			if (m->recursive) {
+				m->owner = thread_get_id();
```
Comment (etienne-lms) [inline core/arch/arm/kernel/mutex.c:61]: `assert(m->owner == THREAD_ID_INVALID);` for debug?

File: core/arch/arm/include/kernel/mutex.h [hunk 2]
```diff
@@ -7,18 +7,26 @@
 
 #include <types_ext.h>
 #include <sys/queue.h>
+#include <kernel/refcount.h>
```
Comment (etienne-lms) [inline core/arch/arm/include/kernel/mutex.h:10]: minor: maybe move this line and the below one 2 lines above now.

minor: I suggest you update the copyright date here and in **mutex.c**.

**Comment by jbech-linaro** [comment]
IBART seems happy now, so ... that means that I don't have to spend time on trying this out I suppose :)

File: core/arch/arm/include/kernel/mutex.h (hunk 2 above)
Comment (jforissier) [reply core/arch/arm/include/kernel/mutex.h:8]: > minor: I suggest you update the copyright date here and in **mutex.c**.

I'm not sure I can simply change the dates, I think I would have to add a Huawei copyright instead (I'm not contributing as Linaro currently). Does it really matter? Honestly I don't know.

**Comment by jforissier** [comment]
> IBART seems happy now, so ... that means that I don't have to spend time on trying this out I suppose :)

This time it didn't time out indeed, it just printed the (erroneous) "failed" message. But IBART will likely remain a bit unreliable until it uses https://github.com/96boards-hikey/burn-boot/pull/4 I'm afraid.

**Comment by jenswi-linaro** [comment]
I'd prefer a new mutex type, `struct mutex_recursive` for instance. It's bit about conserving memory but more to avoid changing the semantics of a `struct mutex`.

**Comment by jbech-linaro** [comment]
@jforissier 

```bash
$ rpo_rev 
OpenPlatformPkg -- fbdd4ae Merge pull request #83 from hzhuang1/rm_earlycon                                                          
atf-fastboot -- d75cfac Merge pull request #10 from hzhuang1/tbb
build -- e388a72 qemu: update u-boot config fragment for v2019.01 to v2020.04                                                        
buildroot -- 95942f5fcd boot/optee-os: new package
burn-boot -- f0cfabd hisi-idt.py: remove unused function startterm()                                                                 
busybox -- 7ab00a0de Bump version to 1.24.0
edk2 -- 77326b5a15 Merge pull request #73 from rsalveti/hikey960_v2.5                                                                
grub -- e54c99aaf Increase version to 2.02.
l-loader -- 49db0a0 Merge pull request #28 from ricardosalveti/hikey960_v1.2                                                         
linux -- 5973cd5de376 ARM: dts: stm32: Define OP-TEE resources on stm32mp15 and enable on DK2                                        
optee_client -- c77b4aa flags: do not override CFLAGS from host
optee_examples -- a66caa4 hotp: fix buffer overflow issue
optee_os -- 840dd114 [Review] core: introduce recursive mutexes
optee_test -- 391168e regression 1023: workaround Clang optimization issue                                                           
patches_hikey -- d9c07f0 grub.configfile: Remove unnecessary fs_label                                                                
trusted-firmware-a -- a04808c16 Merge "Update TF-A version to 2.2" into integration                      
```

All fine, build, flash, boot, xtest ...
```basb
+-----------------------------------------------------
24790 subtests of which 0 failed
103 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
```
`Tested-by: Joakim Bech <joakim.bech@linaro.org>` (HiKey6220)

**Comment by jforissier** [comment]
> I'd prefer a new mutex type, `struct mutex_recursive` for instance. It's bit about conserving memory but more to avoid changing the semantics of a `struct mutex`.

Doable of course, but how would you avoid adding and duplicating too much code? That is, if we use the same debugging scheme as regular mutexes, we would be adding (mostly duplicated code):
```
mutex_recursive_lock()
mutex_recursive_lock_debug()
__mutex_recursive_lock()

mutex_recursive_unlock()
mutex_recursive_unlock_debug()
__mutex_recursive_unlock()
```

What's wrong with extending the semantics of `struct mutex`? "Recursive" is just an attribute to indicate which kind of behavior we want in one particular case. It does not change the validity of any operation (lock, unlock, trylock, read_lock, read_trylock).
Also, it is somewhat similar to pthread mutexes (where "recursive" is an attribute).

**Comment by jenswi-linaro** [comment]
You could declare something like:
``` C
struct mutex_recursive {
         struct mutex m;
         bool recursive;
         int owner;
         struct refcount lock_count
};
```
That way you can use the normal mutex functions when needed. I realize that `__mutex_lock()` and `__mutex_unlock()` might need an `struct mutex_recursive *mr` argument to handle recursive mutexes but that's not so bad as it's just an internal implementation detail.

We may want to change `int owner` into `short int owner` to save a few bytes, if needed we can add a compile time assert that the number of threads cannot exceed `SHRT_MAX`.

**Comment by jforissier** [comment]
@jenswi-linaro https://github.com/OP-TEE/optee_os/pull/3969

**Pull request closed without merging**